### PR TITLE
Three features: SCP, Merge with joinkey, Secret with local encryption

### PR DESCRIPTION
**Pull request by haihongren**
**Added SCP API feature:** 

- SCP feature allows user to retrieve a JSON file from remote machine(s) and parse JSON payload using Flex. It uses SSH as the underlying mechanism for secure communication. 

- SCP API supports following parameters
```
- User    
- Pass    -  treated as encrypted password if Passphrase is not empty
- Host     
- Port      
- RemoteFile - path to file on the remote server(e.g. /meta/cmdb.json), assumed to be a JSON file
- Passphrase - passprhase used to decrypt the password
- SSHPEMFile - if provided, pem file will be used for authentication instead of user/pass.
```
- added the following two parameters in Global config struct

```
- Passphrase - if present, this will be used, local Passphrase will override global.
- SSHPEMFile - if present, this will be used, local SSHPEMFile will override global.
```
**Added Merge with Joinkey feature:** 
- Merge with joinkey feature allows the results from multiple APIs to be joint and merged into a new event. 
- The first API with `merge` and `joinkey` parameter is the primary event, the rest will be treated as secondary event.
- The joinkey(s) in primary event is used to lookup secondary event(s) on the joinkey specified, if a value match is found, the attributes from the secondary event(s) will be merged into the primary event, in another word, the primary event is enriched using secondary event(s).
- For API with only `merge` parameter, it does merge operation only as before.  Merge expects a single sample per event.  
- The follow Flex example config yml demonstrates a Merge with joinkey scenario.

```
name: mergeJoinSample
global:
base_url: http://127.0.0.1:3000/
headers:
  X-Vault-Token: myroot
apis:
- event_type: usageInfo
  url: usageInfo
  merge: encrichedUsageInfo
  joinkey: priceId,customerId,departmentId

- event_type: priceDetails
  url: priceDetails
  merge: encrichedUsageInfo
  joinkey: priceId

- event_type: customDetails
  url: customDetails
  merge: encrichedUsageInfo
  joinkey: customerId

- event_type: departmentDetails
  url: departmentDetails
  merge: encrichedUsageInfo
  joinkey: departmentId

```

**Added `local` encryption type support for Secret Management** 
- The following Flex yml example shows the "kind:local" encryption type. 
```
secrets:
  mylogin: # specify a namespace for your key
    kind: local # set kind of secret
    key: n3wr3lic!
    data: c21036ecac6c2592090153c5d4b220ef4e28a6d2bbe6d9c0c3e004194fc6c85137b43358cff8520b1586
    type: value
```
- The `data` contains the encrypted data, which is generated by running nri-flex with the following arguments. 
`nri-flex -encrypt_pass 's3cr3tPassw0rd' -pass_phrase 'n3wr3lic!'`
-  The parameters follow the same convention used in Secret Management. 
   For example, 
-- `type` can be value, json or equal.
-- The encrypted data can be stored in `file`. This helps to separate the key/passphrase and encrypted data.

### CODE REVIEW

File: internal/inputs/http.go [hunk 1]
```diff
@@ -18,10 +18,10 @@ import (
 	"strings"
 	"time"
 
+	xj "github.com/basgys/goxml2json"
```
Comment (Kav91) [inline internal/inputs/http.go]: @haihongren we need to add this package to this file -> https://github.com/newrelic/nri-flex/blob/master/THIRD_PARTY_NOTICES.md

Comment (haihongren) [reply internal/inputs/http.go]: done!tx

File: examples/flexConfigs/sqlserver-database-example.yml [hunk 2]
```diff
@@ -1,14 +1,12 @@
...
-    custom_attributes: # applies to all queries
-      host: myDbServer
     db_async: true # process queries async
-    db_queries: 
-      - name: sqlserverExecSample
-        run: Select cpu_time,database_id,deadlock_priority,endpoint_id from sys.dm_exec_sessions
-        custom_attributes: # can apply additional at a nested level
-          nestedAttr: nestedVal
\ No newline at end of file
+    db_queries:
```
Comment (Kav91) [inline examples/flexConfigs/sqlserver-database-example.yml]: No need to update this example, lets remove this from the PR.

**Comment by bokken** [comment]
Will the Merge with Joinkey feature be able to work in combination with stored lookups and multiple results?

For example my first command returns 2 events with keys placed into store_lookups.
The second and third command both lookup those keys and each produce 2 results which should be merged together.

So the first command is `gluster volume info` which produces n events unique identified by "volumeName":

```
volumeName: partition000
...
status: Started

volumeName: partition001
...
status: Started
```

The second command is `gluster volume status ${lookup:volumeName} detail` which produces n events for each volume unique identified by "brick":

```
Brick:dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster1
...
Online:Y

Brick:dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster2
...
Online:Y
```

The third command is `gluster volume heal ${lookup:volumeName} statistics heal-count` which produces the same n events for each volume uniquely identified by "brick":

```
Brick:dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster1
Number of entries: 0

Brick:dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster2
Number of entries: 0
```

The desired result would be to have the "Number of entries" key/value merged into the events from the second command.

**Comment by haihongren** [comment]
> Will the Merge with Joinkey feature be able to work in combination with stored lookups and multiple results?
> 
> For example my first command returns 2 events with keys placed into store_lookups.
> The second and third command both lookup those keys and each produce 2 results which should be merged together.
> 
> So the first command is `gluster volume info` which produces n events unique identified by "volumeName":
> 
> ```
> volumeName: partition000
> ...
> status: Started
> 
> volumeName: partition001
> ...
> status: Started
> ```
> 
> The second command is `gluster volume status ${lookup:volumeName} detail` which produces n events for each volume unique identified by "brick":
> 
> ```
> Brick:dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster1
> ...
> Online:Y
> 
> Brick:dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster2
> ...
> Online:Y
> ```
> 
> The third command is `gluster volume heal ${lookup:volumeName} statistics heal-count` which produces the same n events for each volume uniquely identified by "brick":
> 
> ```
> Brick:dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster1
> Number of entries: 0
> 
> Brick:dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster2
> Number of entries: 0
> ```
> 
> The desired result would be to have the "Number of entries" key/value merged into the events from the second command.

Can you please send me the yml file and the sample output events? 

./nri-flex -config_file <your config yml file> -pretty

**Comment by bokken** [comment]
```
name: glusterVolumes
apis:
  - name: glusterVolumes
    commands:
      - run: gluster volume info | tail -n +2
        split_by: ": "
        split_output: ^\s*$ # regex to split by line
    convert_space: "_"
    snake_to_camel: true
    remove_keys:
      - ^brick.*$
    store_lookups:
      volumeName: volumeName
  - name: glusterPartitions
    commands:
        #strip first 2 lines; remove spaces separating keys and values; remove trailing spaces; remove word "Brick " in front of actual brick name
      - run: gluster volume status ${lookup:volumeName} detail | tail -n +3 | sed 's|\s\{1,\}:\s|:|g' | sed 's|\s\{1,\}$||g' | sed 's|:Brick |:|g'
        split_by: ":"
        split_output: ^--*$
    convert_space: "_"
    snake_to_camel: true
    custom_attributes:
      volumeName: ${lookup:volumeName}
  - name: healStats
    commands:
      - run: "gluster volume heal ${lookup:volumeName} statistics heal-count | tail -n +3 | sed 's|Brick |Brick: |g'"
        split_by: ": "
        split_output: ^\s*$
    convert_space: "_"
    snake_to_camel: true
    custom_attributes:
      volumeName: ${lookup:volumeName}
```

**Comment by bokken** [comment]
The output of the first 2 commands are described on https://github.com/newrelic/nri-flex/issues/58.
Strangely, it appears that results are not being duplicated here when the 3rd command is added.

The output of the 3rd command is as follows:

```
[root@dwxvalgovgluster1 ~]# gluster volume heal partition000 statistics heal-count | tail -n +3 | sed 's|Brick |Brick: |g'
Brick: dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster1
Number of entries: 0

Brick: dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition000/dwxvalgovgluster2
Number of entries: 0
```

And:

```
[root@dwxvalgovgluster1 ~]# gluster volume heal partition001 statistics heal-count | tail -n +3 | sed 's|Brick |Brick: |g'
Brick: dwxvalgovgluster1:/mount/cps/dwxval_gov_gluster_p/partition001/dwxvalgovgluster1
Number of entries: 0

Brick: dwxvalgovgluster2:/mount/cps/dwxval_gov_gluster_p/partition001/dwxvalgovgluster2
Number of entries: 0
```

**Comment by bokken** [comment]
@haihongren,

That is not quite an accurate comparison.
Your services would need to return something more like:

```
{

  "volume": [
    {
      "volname": "v1"
    },
    {
      "volname": "v2"
    }
  ],

  "partition": [
    {
      "volname": "v1",
      "partition": "p1",
      "partitionsize": 111
    },
    {
      "volname": "v1",
      "partition": "p2",
      "partitionsize": 222
    },
    {
      "volname": "v2",
      "partition": "p3",
      "partitionsize": 333
    },
    {
      "volname": "v2",
      "partition": "p4",
      "partitionsize": 444
    }
  ],

  "partitionHeal": [
    {
      "volname": "v1",
      "partition": "p1",
      "partitionHeal": 555
    },
    {
      "volname": "v1",
      "partition": "p2,
      "partitionHeal": 666
    },
    {
      "volname": "v2",
      "partition": "p3",
      "partitionHeal": 777
    },
    {
      "volname": "v2",
      "partition": "p4",
      "partitionHeal": 888
    }
  ]
}
```

I am not trying to get everything merged up to "volume". Rather, what I need is to get the partitionHeal value merged up with the partitionsize where the volname and partition values match.

**Comment by haihongren** [comment]
@bokken 

Unfortunately, the store_lookups does not work with Merge/Joinkey currently. 

However, the lookup_file feature does work with Merge/Joinkey. Not sure whether this helps your use case. 

**Sample Flex yml file using lookup_file**
```yml
name: glusterVolumes

lookup_file: volume_lookup.json ### location of lookup file

apis:
  - name: partition
    url: http://localhost:3000/partition?volname=${lf:volname}
    add_attribute:
      volpart: ${volname}${partition}
    merge: myResult
    joinkey: volpart
  - name: partitionHeal
    url: http://localhost:3000/partitionHeal?volname=${lf:volname}
    add_attribute:
      volpart: ${volname}${partition}
    merge: myResult
    joinkey: volpart
```

**sample volume_lookup.json**
```json
[{ "volname": "v1" }, { "volname": "v2" }]
```

**sample Flex output**
```json
                        "metrics": [
                                {
                                        "api.StatusCode": 200,
                                        "event_type": "myResult",
                                        "integration_name": "com.newrelic.nri-flex",
                                        "integration_version": "0.9.2-1-gd21506f-dirty",
                                        "partition": "p3",
                                        "partitionHeal": 777,
                                        "partitionsize": 333,
                                        "volname": "v2",
                                        "volpart": "v2p3"
                                },
                                {
                                        "api.StatusCode": 200,
                                        "event_type": "myResult",
                                        "integration_name": "com.newrelic.nri-flex",
                                        "integration_version": "0.9.2-1-gd21506f-dirty",
                                        "partition": "p4",
                                        "partitionHeal": 888,
                                        "partitionsize": 444,
                                        "volname": "v2",
                                        "volpart": "v2p4"
                                },
                                {
                                        "api.StatusCode": 200,
                                        "event_type": "myResult",
                                        "integration_name": "com.newrelic.nri-flex",
                                        "integration_version": "0.9.2-1-gd21506f-dirty",
                                        "partition": "p1",
                                        "partitionHeal": 555,
                                        "partitionsize": 111,
                                        "volname": "v1",
                                        "volpart": "v1p1"
                                },
                                {
                                        "api.StatusCode": 200,
                                        "event_type": "myResult",
                                        "integration_name": "com.newrelic.nri-flex",
                                        "integration_version": "0.9.2-1-gd21506f-dirty",
                                        "partition": "p2",
                                        "partitionHeal": 666,
                                        "partitionsize": 222,
                                        "volname": "v1",
                                        "volpart": "v1p2"
                                },
                                {
                                        "event_type": "flexStatusSample",
                                        "flex.Hostname": "C02W60KWHTD8",
                                        "flex.IntegrationVersion": "0.9.2-1-gd21506f-dirty",
                                        "flex.counter.ConfigsProcessed": 2,
                                        "flex.counter.EventCount": 4,
                                        "flex.counter.EventDropCount": 0,
                                        "flex.counter.HttpRequests": 4,
                                        "flex.counter.myResult": 4,
                                        "flex.time.elaspedMs": 25,
                                        "flex.time.endMs": 1579660247995,
                                        "flex.time.startMs": 1579660247970
                                }
                        ],

**Comment by bokken** [comment]
I logged new request to have Merge/JoinKey to work with stored lookups.
Thanks.

**Pull request merged**